File: 0206-reverse-linked-list/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head){
    if (head == NULL) return NULL;

    struct ListNode * solution = head;
    if (head -> next) {
        solution = reverseList(head -> next);
        head -> next -> next = head;
    }
    head -> next = NULL;
    return solution;
}

File: 0206-reverse-linked-list/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
    var previous *ListNode
    current := head
    for current != nil {
        next_current := current.Next
        current.Next = previous
        previous = current
        current = next_current
    }
    return previous
}
